src/fd.cpp

```diff
@@ -110,19 +159,99 @@ static int wait_thread(void *arg) {
     std::fill(args->results->begin(), args->results->end(), NA_INTEGER);
   }
 
+  return 0;
+
+}
+
+static int wait_thread_single(void *arg) {
+
+  tct_thrd_detach(tct_thrd_current());
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  if (wait_on_fds(args))
+    return 1;
+
   callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
 
   return 0;
 
 }
 
+static int wait_thread_persistent(void *arg) {
+
+  tct_thrd_detach(tct_thrd_current());
+  thread_active->store(true); // atomic allows update prior to acquiring lock
+
+  if (cv.lock()) THREAD_RETURN(1);
+  if (cv.signal()) { cv.unlock(); THREAD_RETURN(1); } // signal to sync with main thread
+  cv.busy(false);
+  while (!cv.busy()) {
+    if (cv.wait()) { cv.unlock(); THREAD_RETURN(1); }
+  }
+  if (cv.unlock()) THREAD_RETURN(1);
+
+  while (1) {
+
+    if (!thread_active->load()) break; // set by cv destructor on exit
+
+    std::shared_ptr<ThreadArgs> args = *thread_args;
+
+    if (wait_on_fds(args)) {
+      thread_args.reset(); // cancelled so just reset
+    } else {
+      callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(thread_args.release()), 0, args->loop);
+    }
+
+    if (cv.lock()) THREAD_RETURN(1);
+    cv.busy(false);
+    while (!cv.busy()) {
+      if (cv.wait()) { cv.unlock(); THREAD_RETURN(1); }
+    }
+    if (cv.unlock()) THREAD_RETURN(1);
+
+  }
+
+  cv.destroy(); // must do this from here rather than cv destructor to avoid deadlock
+  return 0;
+
+}
+
 static int execLater_launch_thread(std::shared_ptr<ThreadArgs> args) {
 
   std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(new std::shared_ptr<ThreadArgs>(args));
 
+  if (!thread_active->load()) {
+
+    tct_thrd_t thr;
+    if (tct_thrd_create(&thr, &wait_thread_persistent, NULL) != tct_thrd_success)
+      return 1;
+
+    if (cv.lock()) return 1;
+    while (!thread_active->load()) { // thread can update this without acquiring lock
+      if (cv.wait()) { cv.unlock(); return 1; } // wait for signal from thread
+    }
+    if (cv.unlock()) return 1;
+
+  }
+
+  do {
+
+    if (cv.lock()) return 1;
+    if (cv.busy()) { cv.unlock(); break; } // busy so create new single thread
+
+    cv.busy(true);
+    thread_args = std::move(argsptr);
+    if (cv.signal()) { cv.unlock(); return 1; }
+    if (cv.unlock()) return 1;
```

Comment:
General point about the error codes from these synchronization-related operations: if they fail, we're in really bad shape. Like, maybe it's not safe to continue using this process, or else there was a programming error (like unlocking a mutex that wasn't locked). That's why the ConditionVariable class in threadutils.h just throws a C++ exception. I'd rather use that abstraction, which is harder to use incorrectly.

Reply:
I was probably too naive in just following the API I was given... In fact I checked the tct functions today and they should actually all be `void` returns. The error codes in those cases are indeed all for 'programmer error'. In such a case, I'm not averse to using the existing class as the exceptions are never thrown.